biz.aQute.bndlib/src/aQute/bnd/osgi/PluginsContainer.java

```diff
@@ -47,17 +51,72 @@
 	private static final MethodType		defaultConstructor	= methodType(void.class);
 
 	private final Set<Object>			plugins				= new CopyOnWriteArraySet<>();
+	private final Set<AutoCloseable>	closeablePlugins	= new CopyOnWriteArraySet<>();
+
 	// The following are only mutated during init(), so they don't need to
 	// be concurrent-safe
 	private final Set<String>			missingCommand		= new HashSet<>();
-	private final Set<AutoCloseable>	closeablePlugins	= new HashSet<>();
+	private Workspace					ws;
+	private Processor					processor;
+
+	public interface PluginProvider {
+		Stream<?> provide(Class<?> type);
+	}
+
+	class AbstractPlugin<T> implements PluginProvider {
+		final Class<T>	serviceClass;
+		final List<T>	externals	= new ArrayList<>();
+		final Attrs		attrs;
+		boolean			inited;
+
+		AbstractPlugin(Class<T> type, Attrs attrs) {
+			serviceClass = type;
+			this.attrs = attrs;
+		}
+
+		@SuppressWarnings("unchecked")
+		void init() {
+			if (!inited) {
```

Comment:
You're right but I do not think AtomicBoolean is correct, it needs to be synchronized or you can get duplicates. I recall I had it synchronized then reasoned it wasn't needed, then made changes and now it can be called on any thread.